### PR TITLE
[FLINK-11388][fs] Add Aliyun OSS recoverable writer

### PR DESCRIPTION
**Pull request by wujinhu**
## What is the purpose of the change

OSS offers persistence only after uploads or multi-part uploads complete. In order to make streaming uses OSS as sink, we should implement a Recoverable writer. This writer will snapshot and store multi-part upload information and recover from those information when failure occurs.

## Brief change log

**- add OSS recoverable writer**

## Verifying this change

This implementation adds some tests to test logic of recoverable writer. However, in order to run these tests, someone need to have his own Aliyun access key id and access key secret. Then, set environment variables below：
`export ARTIFACTS_OSS_ENDPOINT=<Aliyun OSS endpoint to connect to>`
`export ARTIFACTS_OSS_BUCKET=<Your bucket>`
`export ARTIFACTS_OSS_ACCESS_KEY=<Your access key id>`
`export ARTIFACTS_OSS_SECRET_KEY=<Your access secret>`

These tests are skipped by default.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by JingsongLi** [comment]
Thanks @wujinhu , I will take a look~

File: flink-filesystems/flink-fs-common/pom.xml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	<modelVersion>4.0.0</modelVersion>
+
+	<parent>
+		<artifactId>flink-filesystems</artifactId>
+		<groupId>org.apache.flink</groupId>
+		<version>1.15-SNAPSHOT</version>
+		<relativePath>..</relativePath>
+	</parent>
+
+	<artifactId>flink-fs-common</artifactId>
```
Comment (JingsongLi) [inline flink-filesystems/flink-fs-common/pom.xml:30]: I think we can just move utils to `flink-core` just like `RefCountedFile`.

**Comment by JingsongLi** [comment]
You can create a JIRA for updating document in: https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/filesystem/

File: flink-filesystems/flink-oss-fs-hadoop/src/main/java/org/apache/flink/fs/osshadoop/OSSAccessor.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                                    + "File length is %d bytes, RecoveryData indicates %d bytes",
+                            contentLength, length));
+        }
+
+        org.apache.hadoop.fs.Path srcPath = new org.apache.hadoop.fs.Path("/" + objectName);
+        org.apache.hadoop.fs.Path localPath = new org.apache.hadoop.fs.Path(dstPath);
+        fs.copyToLocalFile(srcPath, localPath);
+
+        String crcFileName = "." + localPath.getName() + ".crc";
+        (new File(localPath.getParent().toString() + "/" + crcFileName)).delete();
```
Comment (JingsongLi) [inline flink-filesystems/flink-oss-fs-hadoop/src/main/java/org/apache/flink/fs/osshadoop/OSSAccessor.java:98]: Remove `()`?

File: flink-filesystems/flink-fs-common/pom.xml (hunk 1 above)
Comment (wujinhu) [reply flink-filesystems/flink-fs-common/pom.xml:30]: Ok, it seems good

**Comment by wujinhu** [comment]
@flinkbot run azure

**Review by JingsongLi — APPROVED** [review]
Looks good to me!

**Pull request merged**